File: Gemi3/pyCAVI/_update_x.py
import numpy as np
import sys
from Gemi3.pyCAVI.util import find_s_keys,find_r_keys

def x_loop(self):
    if self.verbose:
        print("Updating x values",file=sys.stdout)
    for gi in self.unique_guides:
        indices=collect_all_keys(self,gi)
        numerator_sum=np.sum(np.asarray(list(map(lambda x: x_numerator(self,x),indices))))
        denominator_sum=np.sum(np.asarray(list(map(lambda x: x_denominator(self,x),indices))))
        new_xgi=((self.prio_tau_x*self.prio_mu_x)+numerator_sum)/(self.prio_tau_x+denominator_sum)
        if self.debug:print(f"Old xgi {self.x[self.hash_x[gi]]}.",sys.stdout)
        self.x[self.hash_x[gi]]= new_xgi
        if self.debug: print(f"New xgi {self.x[self.hash_x[gi]]}.", sys.stdout)

def xx_loop(self):
    if self.verbose:
        print("Updating xx values",file=sys.stdout)
    for gi in self.unique_guides:
        indices = collect_all_keys(self, gi)
        numerator_sum = np.sum(np.asarray(list(map(lambda x: x_numerator(self, x), indices))))
        denominator_sum = np.sum(np.asarray(list(map(lambda x: x_denominator(self, x), indices))))
        new_xxgi = ((self.prio_tau_xx * self.prio_mu_xx) + numerator_sum) / (self.prio_tau_xx + denominator_sum)
        if self.debug: print(f"Old xxgi {self.xx[self.hash_x[gi]]}.", sys.stdout)
        self.xx[self.hash_x[gi]] = new_xxgi
        if self.debug: print(f"New xxgi {self.xx[self.hash_x[gi]]}.", sys.stdout)

def x_numerator(self,key_dict):
    numerator=(self.tau[self.hash_t[key_dict["t"]]]*self.y[self.hash_y[key_dict["gl"]]])*\
    (self.D[self.hash_d[key_dict["t"]]]-\
    (self.x[self.hash_x[key_dict["hj"]]]*self.y[self.hash_y[key_dict["hl"]]])-\
    (self.x[self.hash_x[key_dict["fk"]]]*self.y[self.hash_y[key_dict["fl"]]])-\
    (self.r[self.hash_r[key_dict["r"]]]*self.s[self.hash_s[key_dict["s"]]]))
    return numerator

def x_denominator(self,key_dict):
    denominator = self.tau[self.hash_t[key_dict["t"]]] * self.yy[self.hash_y[key_dict["gl"]]]
    return denominator


def collect_all_keys(self,gi_extern):
    #only get keys from tau and split them accordingly
    tau_keys = [x for x in self.hash_t.keys() if gi_extern in x] #TODO check in "if x.startswith(gi)" is needed
    out=[]
    for k in tau_keys:
        gi,hj,fk,l = k.split("_")
        gl = self.guide_to_gene(gi)
        hl = self.guide_to_gene(hj)
        fl = self.guide_to_gene(fk)
        s = find_s_keys(self,gl,hl,fl,l)
        r = find_r_keys(self,gi,hj,fk)
        out.append({"t":k,"gi":gi,"hj":hj,"fk":fk,"l":l,
                    "gl":"_".join([gl,l]),
                    "hl":"_".join([hl,l]),
                    "fl":"_".join([hl,l]),
                    "r":r,
                    "s":s})
    return out

File: src/CAVI/update_x.rs
use super::CAVI;
use rayon::prelude::*;
use std::collections::HashMap;

impl<'a> CAVI<'a> {
    pub fn update_x(&self) -> Result<(), Box<dyn std::error::Error>> {
        if self.verbose {
            println! {"Updating x values"}
        }
        for gi in self.lookup.read().unwrap().get("x").unwrap().keys() {
            let all_numerators: Vec<f64> = self
                .lookup
                .read()
                .unwrap()
                .get("x")
                .unwrap()
                .get(gi)
                .unwrap()
                .par_iter()
                .map(|element| x_numerator(self, element))
                .collect();
            let all_denominators: Vec<f64> = self
                .lookup
                .read()
                .unwrap()
                .get("x")
                .unwrap()
                .get(gi)
                .unwrap()
                .par_iter()
                .map(|element| x_denominator(self, element))
                .collect();
            let new_xgi: f64 = ((self.prio_tau_x * self.prio_mu_x)
                + all_numerators.par_iter().sum::<f64>())
                / (self.prio_tau_x + all_denominators.par_iter().sum::<f64>());
            if self.debug {
                println!(
                    "Old xgi {}.",
                    *self
                        .all_x
                        .read()
                        .unwrap()
                        .get(*self.hash_x.get(*gi).unwrap())
                        .unwrap()
                )
            }
            {
                *self
                    .all_x
                    .write()
                    .unwrap()
                    .get_mut(*self.hash_x.get(*gi).unwrap())
                    .unwrap() = new_xgi
            }
            if self.debug {
                println!(
                    "New xgi {}.",
                    *self
                        .all_x
                        .read()
                        .unwrap()
                        .get(*self.hash_x.get(*gi).unwrap())
                        .unwrap()
                )
            }
        }
        return Ok(());
    }

    pub fn update_xx(&self) -> Result<(), Box<dyn std::error::Error>> {
        if self.verbose {
            println! {"Updating xx values"}
        }
        for gi in self.lookup.read().unwrap().get("x").unwrap().keys() {
            let all_numerators: Vec<f64> = self
                .lookup
                .read()
                .unwrap()
                .get("x")
                .unwrap()
                .get(gi)
                .unwrap()
                .par_iter()
                .map(|element| x_numerator(self, element))
                .collect();
            let all_denominators: Vec<f64> = self
                .lookup
                .read()
                .unwrap()
                .get("x")
                .unwrap()
                .get(gi)
                .unwrap()
                .par_iter()
                .map(|element| x_denominator(self, element))
                .collect();
            let new_xxgi: f64 = ((self.prio_tau_xx * self.prio_mu_xx)
                + all_numerators.par_iter().sum::<f64>())
                / (self.prio_tau_xx + all_denominators.par_iter().sum::<f64>());
            if self.debug {
                println!(
                    "Old xxgi {}.",
                    *self
                        .all_xx
                        .read()
                        .unwrap()
                        .get(*self.hash_x.get(*gi).unwrap())
                        .unwrap()
                )
            }
            {
                *self
                    .all_xx
                    .write()
                    .unwrap()
                    .get_mut(*self.hash_x.get(*gi).unwrap())
                    .unwrap() = new_xxgi
            }
            if self.debug {
                println!(
                    "New xxgi {}.",
                    *self
                        .all_xx
                        .read()
                        .unwrap()
                        .get(*self.hash_x.get(*gi).unwrap())
                        .unwrap()
                )
            }
        }
        return Ok(());
    }
}

fn x_numerator(obj: &CAVI, key_dict: &HashMap<&str, &str>) -> f64 {
    let numerator: f64 = (*obj
        .all_tau
        .read()
        .unwrap()
        .get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap())
        .unwrap()
        * *obj
            .all_y
            .read()
            .unwrap()
            .get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap())
            .unwrap())
        * (*obj
            .all_D
            .read()
            .unwrap()
            .get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap())
            .unwrap()
            - *obj
                .all_x
                .read()
                .unwrap()
                .get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap())
                .unwrap()
                * *obj
                    .all_y
                    .read()
                    .unwrap()
                    .get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap())
                    .unwrap()
            - *obj
                .all_x
                .read()
                .unwrap()
                .get(*obj.hash_x.get(*key_dict.get("fk").unwrap()).unwrap())
                .unwrap()
                * *obj
                    .all_y
                    .read()
                    .unwrap()
                    .get(*obj.hash_y.get(*key_dict.get("fl").unwrap()).unwrap())
                    .unwrap()
            - *obj
                .all_r
                .read()
                .unwrap()
                .get(*obj.hash_r.get(*key_dict.get("r").unwrap()).unwrap())
                .unwrap()
                * *obj
                    .all_s
                    .read()
                    .unwrap()
                    .get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap())
                    .unwrap());
    return numerator;
}

fn x_denominator(obj: &CAVI, key_dict: &HashMap<&str, &str>) -> f64 {
    let denominator: f64 = *obj
        .all_tau
        .read()
        .unwrap()
        .get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap())
        .unwrap()
        * *obj
            .all_yy
            .read()
            .unwrap()
            .get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap())
            .unwrap();
    return denominator;
}
